sklearn/preprocessing/tests/test_function_transformer.py

```diff
@@ -156,15 +158,92 @@ def test_check_inverse():
         Xt = assert_no_warnings(trans.fit_transform, X)
         assert_allclose_dense_sparse(X, trans.inverse_transform(Xt))
 
     # check that we don't check inverse when one of the func or inverse is not
     # provided.
     trans = FunctionTransformer(func=np.expm1, inverse_func=None,
-                                check_inverse=True)
+                                check_inverse=True, validate=True)
     assert_no_warnings(trans.fit, X_dense)
     trans = FunctionTransformer(func=None, inverse_func=np.expm1,
-                                check_inverse=True)
+                                check_inverse=True, validate=True)
     assert_no_warnings(trans.fit, X_dense)
+
+
+@pytest.mark.parametrize(
+    "X",
+    [np.array([[0, 1], [2, 3]]),
+     np.array([[0, 1], [2, np.nan]]),
+     np.array([[0, 1], [2, np.inf]]),
+     np.array([[0, 1], [np.inf, np.nan]])]
+)
+@pytest.mark.parametrize(
+    "force_all_finite",
+    [True, False, 'allow-nan']
+)
+def test_function_transformer_finiteness_pandas(X, force_all_finite):
+    pd = pytest.importorskip('pandas')
+    X_df = pd.DataFrame(X)
+
+    def func(X):
+        return X.columns
+
+    transformer = FunctionTransformer(force_all_finite=force_all_finite,
+                                      validate=True)
+
+    should_fail = False
+    if force_all_finite is True:
+        if not np.isfinite(X).all():
+            should_fail = True
+    elif force_all_finite == 'allow-nan':
+        if np.isinf(X).any():
+            should_fail = True
+
+    if should_fail:
+        with pytest.raises(ValueError, match="Input contains"):
+            transformer.fit_transform(X_df)
+    else:
+        transformer.fit_transform(X_df)
+
+
+@pytest.mark.parametrize(
+    "is_dataframe",
+    [True, False]
+)
+def test_function_transformer_future_warning(is_dataframe):
+    # FIXME: to be removed in 0.22
+    X = np.random.randn(100, 10)
+    transformer = FunctionTransformer()
+    if is_dataframe:
+        pd = pytest.importorskip('pandas')
+        X_df = pd.DataFrame(X)
+        with pytest.warns(FutureWarning):
+            transformer.fit_transform(X_df)
+    else:
+        assert_no_warnings(transformer.fit, X)
+
+
+def test_function_transformer_frame():
+    pd = pytest.importorskip('pandas')
+    X_df = pd.DataFrame(np.random.randn(100, 10))
+    transformer = FunctionTransformer(validate='array-or-frame',
+                                      check_inverse=False)
+    X_df_trans = transformer.fit_transform(X_df)
+    assert hasattr(X_df_trans, 'loc')
+
+
+@pytest.mark.parametrize(
+    "params, msg_err",
+    [({'validate': 'random'}, "'validate' should be"),
+     ({'validate': True, 'force_all_finite': 'random'},
+      "'force_all_finite' should be")]
+)
+def test_function_transformer_params_errors(params, msg_err):
+    X = np.random.randn(100, 10)
+    transformer = FunctionTransformer()
+    transformer.set_params(**params)
+    with pytest.raises(ValueError, match=msg_err):
```

Comment:
This is very nice and so much better than the approach proposed in the [second example of Pytest doc](https://docs.pytest.org/en/latest/assert.html#assertions-about-expected-exceptions), but this only works for pytest 3.10 (released less than a year ago), before it ignores the `match` argument (tested with 3.07) producing false positives.

As long as CI used an up to date pytest this should not be an issue, but if we use it it would be worth specifying minimum pytest requirements in the dev documentation.